### ISSUE TITLE
反向传播时梯度出问题

### ISSUE THREAD

**Issue by btobab** [PaddlePaddle/Paddle#34018]
![image](https://user-images.githubusercontent.com/66876429/124760323-36424080-df63-11eb-8805-7456c79e63e9.png)
这是一个简单的VAE，之前我是直接使用了L2范数计算损失，但效果不太好。尝试换成BCE，却出现了类似于梯度截断的问题

**Comment by tink2123** [comment]
dygraph_tmp_453 这个变量没有梯度节点，建议找到这个变量，看下前后是不是有设置了stop_gradient=True

**Comment by btobab** [comment]
但是我尝试把这个变量打印出来，但显示变量没定义，我感觉是反向传播时才会定义

**Comment by btobab** [comment]
唔，就是L2运行没问题，但是换成BCE就出现了上述问题

**Comment by btobab** [comment]
然后报错也很难懂

**Comment by Steffy-zxf** [comment]
@IamDestiny 你好，是否方便给出最小复现代码及环境（paddle版本，Linux or windows) ？

**Comment by btobab** [comment]
好的，环境为AI Studio，paddle版本2.0.2及2.1.0都试过了，但是都报同样的错
[项目链接](https://aistudio.baidu.com/aistudio/projectdetail/1950873)

**Comment by Steffy-zxf** [comment]
@IamDestiny 麻烦给一个最小的复现代码，运行项目发现在展示图片时出现错误，与以上提及错误无关。麻烦给出最小复现代码，方便排查问题。
![image](https://user-images.githubusercontent.com/48793257/125882383-3c0ee54d-2e0f-4f5e-b422-00aa9d53b47b.png)

**Comment by btobab** [comment]
好哒好哒
[layer.py.zip](https://github.com/PaddlePaddle/Paddle/files/6827836/layer.py.zip)
```
from layer import ConvBN
class Auto_encoder(nn.Layer):
    def __init__(self):
        super(Auto_encoder, self).__init__()
        self.encoder = nn.Sequential(
            ConvBN(3, 64, 5, [2, 2]),
            ConvBN(64, 128, 5, [2, 2]),
            ConvBN(128, 256, 5, [1, 2]),
            ConvBN(256, 512, 5, [1, 1]),
            nn.Conv2D(512, 3, [5, 1], [1, 1])
        )
        self.decoder = nn.Sequential(
            ConvBN(3, 256, 1, 1),
            nn.PixelShuffle(16),
            ConvBN(1, 64, 5, 3),
            ConvBN(64, 256, 5, 1),
            ConvBN(256, 512, 8, 1),
            ConvBN(512, 3, 9, 1)
        )

    def forward(self, x):
        y = self.encoder(x)
        zeros = paddle.zeros([y.shape[0], 3, 60, 30])
        y = paddle.concat([y, zeros], axis=-1)
        y = self.decoder(y)
        return y


model = Auto_encoder()
x = paddle.randn([1, 3, 300, 300])
y = model(x)
print(y.shape)

num_epoch = 35
learning_rate = 3e-3
scheduler = paddle.optimizer.lr.CosineAnnealingDecay(learning_rate=learning_rate, T_max=3, eta_min=learning_rate / 10)
opt = paddle.optimizer.Momentum(scheduler, parameters=model.parameters())
Loss = nn.BCELoss()
for epoch in range(num_epoch):
    for i, data in enumerate(dataloader()):
        data = data.astype("float32")
        data_hat = model(data)
        loss = paddle.mean(F.binary_cross_entropy(data, data_hat))
        loss.backward()
        opt.step()
        opt.clear_grad()
        if i % 100 == 0:
            print("epoch:%d,i:%d,loss:%f" % (epoch, i, loss.numpy()))
    show_imgs([data_hat[0], data_hat[1], data_hat[2]])

    paddle.save(model.state_dict(), path="work/auto.pdparams")
```

**Comment by GT-ZhangAcer** [comment]
## 补充一份容易复现的代码
```
import paddle
import paddle.nn as nn
import paddle.nn.functional as F

class ConvBN(nn.Layer):
    def __init__(self, channels, num_filter, filter_size, stride, padding="valid"):
        super(ConvBN, self).__init__()
        self.conv = nn.Conv2D(channels, num_filter, filter_size, stride=stride, padding=padding)
        self.bn = nn.BatchNorm(num_filter)

    def mish(self, x):
        return x * F.tanh(F.softplus(x))

    def forward(self, x):
        x = self.conv(x)
        x = self.bn(x)
        x = self.mish(x)
        return x

class Auto_encoder(nn.Layer):
    def __init__(self):
        super(Auto_encoder, self).__init__()
        self.encoder = nn.Sequential(
            ConvBN(3, 64, 5, [2, 2]),
            ConvBN(64, 128, 5, [2, 2]),
            ConvBN(128, 256, 5, [1, 2]),
            ConvBN(256, 512, 5, [1, 1]),
            nn.Conv2D(512, 3, [5, 1], [1, 1])
        )
        self.decoder = nn.Sequential(
            ConvBN(3, 256, 1, 1),
            nn.PixelShuffle(16),
            ConvBN(1, 64, 5, 3),
            ConvBN(64, 256, 5, 1),
            ConvBN(256, 512, 8, 1),
            ConvBN(512, 3, 9, 1)
        )

    def forward(self, x):
        y = self.encoder(x)
        zeros = paddle.zeros([y.shape[0], 3, 60, 30])
        y = paddle.concat([y, zeros], axis=-1)
        y = self.decoder(y)
        y = F.sigmoid(y)
        return y


model = Auto_encoder()

num_epoch = 3
learning_rate = 3e-3
scheduler = paddle.optimizer.lr.CosineAnnealingDecay(learning_rate=learning_rate, T_max=3, eta_min=learning_rate / 10)
opt = paddle.optimizer.Momentum(scheduler, parameters=model.parameters())
Loss = nn.BCELoss()
for epoch in range(num_epoch):
    data = paddle.zeros([1, 3, 300, 300])
    data_hat = model(data)
    # loss = F.mse_loss(data, data_hat) # 正常
    loss = F.binary_cross_entropy(data, data_hat) # 报错
    loss.backward()
    opt.step()
    opt.clear_grad()
    print(loss.numpy())

```

**Comment by Steffy-zxf** [comment]
@IamDestiny 你好！看了你的测试代码，发现是你的用法用问题，`paddle.nn.BCELoss()`的输入第一参数是经过模型网络的输出预测值，第二参数是label。你上面的代码正好相反了

> loss = F.binary_cross_entropy(data, data_hat) # 报错

应该为

> loss = F.binary_cross_entropy(data_hat, data) 

参考：https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/BCELoss_cn.html#bceloss

**Comment by btobab** [comment]
好哒好哒，谢谢姐姐
